plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,203 @@
+package com.mapbox.maps.plugin.viewport.experimental
+
+import com.mapbox.android.gestures.MoveGestureDetector
+import com.mapbox.maps.plugin.delegates.MapDelegateProvider
+import com.mapbox.maps.plugin.gestures.GesturesPlugin
+import com.mapbox.maps.plugin.gestures.OnMoveListener
+import com.mapbox.maps.plugin.gestures.gestures
+import com.mapbox.maps.plugin.viewport.VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+import com.mapbox.maps.plugin.viewport.experimental.data.DefaultViewportTransitionOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.FollowingViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.data.OverviewViewportStateOptions
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.MapboxFollowingState
+import com.mapbox.maps.plugin.viewport.experimental.state.MapboxOverviewState
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportState
+import com.mapbox.maps.plugin.viewport.experimental.transition.DefaultViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.ImmediateViewportTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransition
+import java.lang.IllegalStateException
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStatusObservers = CopyOnWriteArrayList<ViewportStatusObserver>()
+  private val internalStates = CopyOnWriteArraySet<ViewportState>()
+  private var currentCancelable: Cancelable? = null
+  private var currentState: ViewportState? = null
+  private val transitions = HashMap<Pair<ViewportState?, ViewportState?>, ViewportTransition>()
+  private lateinit var delegateProvider: MapDelegateProvider
+  private lateinit var gesturesPlugin: GesturesPlugin
+  private val onMoveListener = object : OnMoveListener {
+    override fun onMoveBegin(detector: MoveGestureDetector) {
+      if (transitionsToIdleUponUserInteraction) {
+        transitionToIdle()
+      }
+    }
+
+    override fun onMove(detector: MoveGestureDetector): Boolean {
+      // no-ops
+      return false
+    }
+
+    override fun onMoveEnd(detector: MoveGestureDetector) {
+      // no-ops
+    }
+
+  }
+
+  override val states: Set<ViewportState>
+    get() = internalStates.toSet()
+
+  override var status: ViewportStatus? = null
+    private set(value) {
+      val previousStatus = field
+      field = value
+      notifyStateChanged(previousStatus, field, VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC)
```

Comment:
should we call `notify` if new value is the same as before?

Reply:
If user transition to the same state, we have a guard logic to ignore such request, so it wouldn't end up setting the same status multiple times.